**Import all the dependencies**

In [54]:
import numpy as np
from collections import Counter

**Note**: There is no use of scikit-learn's KNN model.

However, I used scikit-learn to load the iris datasets and to split datasets into training and testing.

In [55]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

**Defining the euclidean distance function**

In [56]:
def euclidean_distance(x1, x2):
    distances = np.sqrt(np.sum((x1-x2)**2))
    return distances

**Defining the K-nearest neighbors class**

In [57]:
class KNN:

    def __init__(self, k=3):
        self.k = k

    
    def fit(self, X, y):
        self.X_train = X
        self.y_train = y

    def predict(self, X):
        predictions = [self._predict(x) for x in X]
        return predictions
    
    def _predict(self, x):
        # computer the distance
        distances = [euclidean_distance(x, x_train) for x_train in self.X_train]

        # get the closest k
        k_indices = np.argsort(distances)[:self.k]
        k_nearest_labels = [self.y_train[i] for i in k_indices]

        # majority voting
        most_common = Counter(k_nearest_labels).most_common()
        return most_common[0][0]


In [58]:
# Load the iris datasets
iris = load_iris()
X, y = iris.data, iris.target

In [63]:
# Splitting the datasets into the training and testing

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=4321)

In [64]:
# Now create an object of KNN and initialize with the apropriate k
knn = KNN(k=5)

# Fit and predict
knn.fit(X_train, y_train)
predictions = knn.predict(X_test)


In [65]:
print(predictions)

[0, 0, 2, 2, 0, 2, 0, 2, 2, 1, 1, 0, 2, 1, 2, 1, 1, 2, 0, 0, 1, 2, 2, 1, 0, 2, 2, 2, 0, 0]


In [66]:
# Now find the accuracy of our model
accuracy_score = np.sum(predictions == y_test)/len(y_test)
accuracy_score

0.9333333333333333

**Conclusion**: We find that accuracy is bit high , which shows that predicted data is 93.33% true of actual data.